serialization/ksp_plugin.proto

```diff
@@ -0,0 +1,88 @@
+syntax = "proto2";
+
+import "serialization/geometry.proto";
+import "serialization/physics.proto";
+import "serialization/quantities.proto";
+
+package principia.serialization;
+
+message HistoryAndProlongation {
+  required Trajectory history = 1;
+  required Trajectory.Pointer prolongation = 2;
+}
+
+message Vessel {
+  required MasslessBody body = 1;
+  oneof trajectory_bundle {
+    HistoryAndProlongation history_and_prolongation = 2;
+    Trajectory owned_prolongation = 3;
+  }
+}
+
+message Celestial {
+  required MassiveBody body = 1;
+  required HistoryAndProlongation history_and_prolongation = 2;
+}
+
+message Part {
+  required Pair degrees_of_freedom = 1;
+  required Quantity mass = 2;
+  required Multivector gravitational_acceleration_to_be_applied_by_ksp = 3;
+}
+
+message PhysicsBubble {
+  message FullState {
+    message PartIdToPart {
+      required fixed32 part_id = 1;
+      required Part part = 2;
+    }
+    message VesselPointerToDegreesOfFreedom {
+      required Plugin.VesselPointer vessel = 1;
+      required Pair degrees_of_freedom = 2;
+    }
+    repeated Plugin.VesselPointer vessels = 1;
+    repeated PartIdToPart parts = 2;
+    required Pair centre_of_mass = 3;
+    required Trajectory centre_of_mass_trajectory = 4;
+    repeated VesselPointerToDegreesOfFreedom from_centre_of_mass = 5;
+    required Pair correction = 6;
+  }
+  required MasslessBody body = 1;
+  optional FullState current = 2;
+}
+
+message Plugin {
+  // NOTE(egg): Assuming 32-bit. If we have 2^32 vessels we'll probably
+  // encounter other issues on the way.
+  message VesselPointer {
+    required uint32 distance_from_vessels_begin = 1;
+  }
+  message CelestialPointer {
+    required uint32 distance_from_celestials_begin = 1;
+  }
+  message VesselPointerToParent {
+    required VesselPointer vessel = 1;
+    required CelestialPointer parent = 2;
+  }
+  message CelestialPointerToParent {
+    required CelestialPointer celestial = 1;
+    optional CelestialPointer parent = 2;
+  }
```

Comment:
As a matter of fact all of this is complicated because it follows too closely the C++ code.  For instance, I would represent the parent/child relationship by having an `required int32 parent` in GuidToVessel and IndexToCelestial, where that field is an index in the celestial array.  Similarly I would represent dirtyness by having a bool in GuidToVessel.